plugins/mediatek-scaler/fu-mediatek-scaler-device.c

```diff
@@ -109,7 +109,11 @@ fu_mediatek_scaler_device_set_i2c_dev(FuMediatekScalerDevice *self,
 		self->i2c_dev = g_object_ref(g_ptr_array_index(i2c_devs, 0));
 		g_debug("found I2C bus at %s, using this device",
 			fu_udev_device_get_sysfs_path(self->i2c_dev));
-		return fu_udev_device_set_physical_id(self->i2c_dev, "i2c", error);
+
+		/* set device open flags */
+		fu_udev_device_add_open_flag(self->i2c_dev, FU_IO_CHANNEL_OPEN_FLAG_READ);
+		fu_udev_device_add_open_flag(self->i2c_dev, FU_IO_CHANNEL_OPEN_FLAG_WRITE);
+		return fu_udev_device_set_physical_id(self->i2c_dev, "i2c-dev", error);
```

Comment:
are you sure about this? I know some distros have i2c-dev either built-in or manually modprobed at startup, but not all of them -- e.g. Fedora and Arch don't have this.

Reply:
On the system with this chip the `i2c-dev` is automatically probed as below udev output.  It's a Ubuntu system, am not sure other distro.

```
P: /devices/pci0000:00/0000:00:02.0/drm/card0/card0-DP-2/i2c-13/i2c-dev/i2c-13
M: i2c-13
R: 13
U: i2c-dev
D: c 89:13
N: i2c-13
L: 0
E: DEVPATH=/devices/pci0000:00/0000:00:02.0/drm/card0/card0-DP-2/i2c-13/i2c-dev/i2c-13
E: SUBSYSTEM=i2c-dev
E: DEVNAME=/dev/i2c-13
E: MAJOR=89
E: MINOR=13
E: USEC_INITIALIZED=4566075
E: ID_PATH=pci-0000:00:02.0
E: ID_PATH_TAG=pci-0000_00_02_0
E: ID_FOR_SEAT=i2c-dev-pci-0000_00_02_0
E: TAGS=:uaccess:seat:
E: CURRENT_TAGS=:uaccess:seat:
```

Reply:
I think this is quite fragile tbh. What does `cat /sys/class/i2c-adapter/i2c-13/name` say? then you can just match on the correct i2c device rather than doing a little dance in mediatek_scaler (matching the dpaux and then using `fu_udev_device_get_children_with_subsystem` and `fu_udev_device_get_siblings_with_subsystem`) and storing the i2c device as a kind of proxy.

This is the same kind of problem I was trying to fix in https://github.com/fwupd/fwupd/pull/7542 -- maybe it actually works better for you so that we only probe the i2c device if the system _hwid_ matches?

Reply:
name for the i2c-adapter:
```
$ cat /sys/class/i2c-adapter/i2c-13/name
AUX USBC4/DDI TC4/PHY E
```
below is the resulted `quirk id`:
```
I2C\NAME_AUX-USBC4-DDI-TC4-PHY-E&HWID_e9d9375d-755c-548f-9098-f187d7c78449
```

I pushed another commit 78f4ccd4 to set the physical id.

Reply:
> I2C\NAME_AUX-USBC4-DDI-TC4-PHY-E&HWID_e9d9375d-755c-548f-9098-f187d7c78449

I think you should use this -- just like we did with https://github.com/fwupd/fwupd/commit/8388ac45420000b0cdf203a759792e423373ff21

Give me a couple of hours and I'll I'll push a branch that might work.

Reply:
The challenge with the `name` of an `i2c-adapter` and `HWID` is that both information aren't transparent to a normal maintainer to extend plugin on the new systems.

Adding system PIDs in the `quirk` file is relative easier to the maintainers in contrast to reading through fwupd verbose output and choosing the right quirk id for the i2c device of desire.

I'm not sure the `name` whether if it is a constant value all the times or it might be changed (either one of hardware, BIOS, or graphic driver) some time in the future.

Reply:
> is that both information aren't transparent to a normal maintainer

What we have now (e.g. https://github.com/fwupd/fwupd/blob/main/plugins/elantp/elantp.quirk#L9 -- to give the other offender) is a hack -- i.e. if the hwid matches then set a _plugin_ flag which then allows a device to probe -- that's something I want to fix soon too. In general if we're relying on the `i2c-dev` module (that might not exist) to link together things as siblings and children of siblings then I think it's super fragile. If there was a `DEVNAME_I2C` in something like `/sys/class/drm_dp_aux_dev/drm_dp_aux1/uevent` that would be a different argument (and a refcounted proxy...), but that's not something the kernel provides.

Finding a device A and then swapping out the internal device (e.g. https://github.com/fwupd/fwupd/blob/main/plugins/elantp/fu-elantp-i2c-device.c#L128) or setting up either an adhoc, or implied proxy device to use as device B makes it really hard to do any proper lifecycling -- which means when we're saying "wait for the device to come back after resetting it" we need to be sure we're waiting for _device A_ -- or maybe _device B_ -- or maybe _both_ -- and then also set up the internal-device swapping again at precisely the correct time. It's madness to try and debug that. 

> I'm not sure the name whether if it is a constant value all the times or it might be changed

I think it's the closest we have to an kernel ABI for i2c devices.

Reply:
> I think it's the closest we have to an kernel ABI for i2c devices.

Do we have a HWID that is just vendor and SKU? Maybe that would work better for Dell system using this.

Reply:
The other thing we could do of course is match on the `FuDpauxDevice` and then match on the OUI like other MST plugins do. We'd get the i2c proxy from the `ddc` link in the device (e.g. `/sys/devices/pci0000:00/0000:00:02.0/drm/card1/card1-eDP-1/ddc`) -- but I'm not sure that's *always* the i2c device we want.

Reply:
In fact, in the parade_lspcon case it's a completely different i2c device.

Reply:
I think it isn't easy to determine the name for `quirk`.

It requires tool to query the `DDC` capability among the i2c-dev candidates to confirm the right target.  The `name` needs to be encoded by fwupd unlike what is seen from sysfs, and the `HWID` too requires a physical system to compute the `manufacturer` plus `product_sku`.

So a system maker (e.g.: ODM) who wants to adopt this plugin needs to understand plugin details to find out the IDs by themselves, or send every single SKU to the plugin maintainer to make up a comprehensive `quirk` list. Any idea to make the maintainance work more simpler?

Reply:
So, maybe it makes sense to add something like this for quirk-only matches for **all** I2c devices?

	udev_parent = FU_UDEV_DEVICE(fu_device_get_backend_parent_with_subsystem(device, "pci", NULL));
	fu_device_add_instance_u16(device, "VID", fu_udev_device_get_subsystem_vendor(udev_parent));
	fu_device_add_instance_u16(device, "PID", fu_udev_device_get_subsystem_model(udev_parent));
	fu_device_build_instance_id_full(FU_DEVICE(self),
					 FU_DEVICE_INSTANCE_FLAG_QUIRKS,
					 NULL,
					 "PCI",
					 "VID",
					 "PID",
					 NULL))

i.e. move the `probe-vcp` bit unconditionally to the FuI2cDevice? It might be a bit weird to see an i2c device with a:

    [PCI\VID_1028&PID_0C88]
    Plugin = mediatek_scaler

...but it's the instance ID you're using already. @superm1 insane?

Reply:
Let me ask a fundamental question - can we get them to add a DPCD OUI?

If this generation is done and done fine not this one we quirk it with platform, but if next one can get an OUI in place I think this scales it the best.

Reply:
I don't think the OUI is the be-all-and-end-all -- the ddc i2c device isn't always the one we want to use for programming.

Reply:
> I don't think the OUI is the be-all-and-end-all -- the ddc i2c device isn't always the one we want to use for programming.

But in the case of the mediatek scaler is it not the right one every time?

Reply:
@CragW is the ddc link the "right" i2c device?

Reply:
> [PCI\VID_1028&PID_0C88]
> Plugin = mediatek_scaler
> ...but it's the instance ID you're using already

I added it in a deleted 72e2762a because `dpaux` is definitely a wrong place for the instance id, maybe go to `fu-udev-device.c` as your mentioned.

The VID/PID instance id should be a quirk entry to a plugin of desire, we still need to walk through `i2c-dev` candidates to determine the right customer.


> is the ddc link the "right" i2c device?

Systems known adopted this controller are AIO desktops, external displays may support ddc link too. We used to use https://github.com/fwupd/fwupd/blob/main/plugins/mediatek-scaler/fu-mediatek-scaler-device.c#L463 to confirm the right target.

Reply:
> Systems known adopted this controller are AIO desktops

Any chance of sending me a test system? I'd like to add these kinds of devices to CI to ensure we don't break them in the future -- and it would be really good to understand how exactly they look in sysfs. Any other mediatek-scaler devboard hardware welcome instead.

Reply:
> Any chance of sending me a test system?

Checking.

Reply:
> because dpaux is definitely a wrong place for the instance id

I actually tried doing this on my system, as I have a algoltek devboard also with OUI==0 -- i.e. something like:

	commit a38205e18803e7a1f99f9a39bc5bdaff2ec3d41d
	Author: Richard Hughes <richard@hughsie.com>
	Date:   Wed Aug 14 13:33:25 2024 +0100

	    Fallback to the parent PCI device for dpaux devices with an unset OUI
	    
	    Based on a patch by Crag Wang <crag_wang@dell.com>, many thanks.

	diff --git a/libfwupdplugin/fu-dpaux-device.c b/libfwupdplugin/fu-dpaux-device.c
	index 1a8accb1c..1a5e967f6 100644
	--- a/libfwupdplugin/fu-dpaux-device.c
	+++ b/libfwupdplugin/fu-dpaux-device.c
	@@ -165,6 +165,33 @@ fu_dpaux_device_setup(FuDevice *device, GError **error)
						 "DEVID",
						 NULL);
	 
	+	/* fallback to the parent PCI device */
	+	if (priv->dpcd_ieee_oui == 0) {
	+		g_autoptr(FuDevice) pci_parent = NULL;
	+		GPtrArray *instance_ids;
	+
	+		/* get the parent PCI device */
	+		pci_parent = fu_device_get_backend_parent_with_subsystem(device, "pci", error);
	+		if (pci_parent == NULL) {
	+			g_prefix_error(error, "failed to get unset-OUI PCI fallback: ");
	+			return FALSE;
	+		}
	+		if (!fu_device_probe(pci_parent, error)) {
	+			g_prefix_error(error, "failed to probe unset-OUI PCI fallback: ");
	+			return FALSE;
	+		}
	+
	+		/* add the visible instance IDs as quirk IDs */
	+		instance_ids = fu_device_get_instance_ids(pci_parent);
	+		for (guint i = 0; i < instance_ids->len; i++) {
	+			const gchar *instance_id = g_ptr_array_index(instance_ids, i);
	+			g_debug("adding unset-OUI PCI fallback of %s", instance_id);
	+			fu_device_add_instance_id_full(device,
	+						       instance_id,
	+						       FU_DEVICE_INSTANCE_FLAG_QUIRKS);
	+		}
	+	}
	+
		/* success */
		return TRUE;
	 }

But that just gave me the PCI device of my GPU, i.e. `/sys/devices/pci0000:00/0000:00:02.0/drm/card1/card1-DP-5/drm_dp_aux5` gives me `0000:00:02.0` for the PCI subsystem. It certainly wouldn't make sense to add the same `PCI\VEN_8086&DEV_9BC4` and `PCI\VEN_8086&DEV_9BC4&SUBSYS_17AA22C2` to the three (!) dpaux devices on my system that failed to get the OUI value (one because it was unset, two because there's nothing connected).

So what the mediatek_scaler plugin seems to be doing now, is matching the dpaux device, and for every dpaux device we find, creating a FuMediatekScalerDevice, finding the root display adapter **subsystem** IDs (nothing to do with the dpaux device at all, and also not the VID/PID of the PCI device unlike the instance ID would suggest...) checking against an allow-list in the quirk file, and if matching finding the i2c device for the dpaux device using the `../ic2*` trick and then using the latter to actually do the update. Essentially it's not using the HWID match, it's using the subsystem IDs for the GPU. It's kinda icky, quite inefficient, fragile, and doesn't match how we do it in all the other plugins.

We really need to do something like `hughsie/mediatek-scaler-FuI2cDevice` -- we can't use the dpaux to find the correct i2c because it doesn't have a OUI, and I think using the PCI subsys to find the correct dpaux is fragile as it seems it can match on multiple dpaux devices and then will fail when trying to setup "the wrong one".

Does the mediatek_scaler respond to any DPCD queries? I know we're looking at `branch-device` at `0x500` but I'm not against querying more. That's the only way i think it makes sense to "hang off" the dpaux_device.

Reply:
What I don't get is why the `ddc` link (e.g. `/sys/devices/pci0000:00/0000:00:02.0/drm/card1/card1-DP-2/ddc` isn't the i2c device you want to use for the scaler update. I mean, if it was we'd use the **drm** device, not a dpaux device. I mean, we already parse the EDID for those and it seems to the "monitor" we're actually trying to match rather than the MST device. @crag Would you be able to match the mediatek scaler from the EDID? e.g. I see:

	FuDrmDevice:
	  Guid:                 9091fe58-8712-520c-a30f-5a1d6c131be7 ← DRM\VEN_DEL&DEV_415F ⚠
	  Serial:               91CWTS2
	  Flags:                none
	  Guid[quirk]:          d07d4a5f-22c0-5c05-8739-e3377d6aabb1 ← DRM\TYPE_DRM_CONNECTOR
	  PhysicalId:           card1-DP-2
	  BackendId:            /sys/devices/pci0000:00/0000:00:02.0/drm/card1/card1-DP-2
	  AcquiesceDelay:       2500
	  ParentBackendIds:     /sys/devices/pci0000:00/0000:00:02.0
	  Number:               0x2
	  Subsystem:            drm
	  OpenFlags:            none
	  ConnectorId:          119
	  Enabled:              true
	  State:                connected

if that's possible then we can do a very simple patch of reading the `ddc` i2c device from the `FuDrmDevice` device, creating a refcounted i2c proxy and then using the proxy from FuMediatekScalerDevice.

Reply:
it seems I don't have the `ddc ` device under card on the system with fedora installed, I can see `i2c-dev` in udev that is behaving the same as Ubuntu.

[fedora-udevadm.log](https://github.com/user-attachments/files/16657706/fedora-udevadm.log)
[fedora-get-devices.log](https://github.com/user-attachments/files/16657705/fedora-get-devices.log)